%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
\RequirePackage{hyperref}
\usepackage{graphicx}
%\usepackage[utf8]{inputenc} %unicode support
\usepackage[utf8x]{inputenc} 
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails

\renewcommand{\thesection}{}% Remove section references...
\renewcommand{\thesubsection}{\arabic{subsection}}%... from subsections


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% \def\includegraphic{}
% \def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Methodology}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Streamlining differential exon and 3' UTR usage with diffUTR}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1,aff2},
    email={gerberst@student.ethz.ch}
]{\fnm{Stefan} \snm{Gerber}}
\author[
   addressref={aff2},
    email={gerhard.schratt@hest.ethz.ch}
]{\fnm{Gerhard} \snm{Schratt}}
\author[
   addressref={aff1,aff3,aff4},
   corref={aff1},
   email={pierre-luc.germain@hest.ethz.ch}
]{\inits{PLG}\fnm{Pierre-Luc} \snm{Germain}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{
  \orgname{Group of Computational Neurogenomics, D-HEST Institute for Neurosciences, ETH Zürich},
  \street{Winterthurerstrasse 190},
  \postcode{8057}
  \city{Zürich},
  \cny{Switzerland}
}
\address[id=aff2]{
  \orgname{Lab of Systems Neuroscience, D-HEST Institute for Neurosciences, ETH Zürich},
  \street{Winterthurerstrasse 190},
  \postcode{8057}
  \city{Zürich},
  \cny{Switzerland}
}
\address[id=aff3]{
  \orgname{Lab of Statistical Bioinformatics, DMLS, University of Zürich}, % university, etc
  \street{Winterthurerstrasse 190},
  \postcode{8057}
  \city{Zürich},
  \cny{Switzerland}
}
\address[id=aff4]{
  \orgname{SIB Swiss Institute of Bioinformatics}, % university, etc
  \city{Zürich},
  \cny{Switzerland}
}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background} %if any
Despite the importance of alternative poly-adenylation and 3' UTR length for a variety of biological phenomena, there are limited means of detecting UTR changes from standard transcriptomic data.
\parttitle{Results} %if any
We present the \textit{diffUTR} Bioconductor package which streamlines and improves upon differential exon usage (DEU) analyses, and leverages existing DEU tools and alternative poly-adenylation site databases to enable differential 3' UTR usage analysis. We demonstrate the \textit{diffUTR} features and show that it is more flexible and more accurate than state-of-the-art alternatives, both in simulations and in real data.
\parttitle{Conclusions} %if any
\textit{diffUTR} enables differential 3' UTR analysis and more generally facilitates DEU and the exploration of their results.
\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{gene expression}
\kwd{transcriptomic}
\kwd{RNAseq}
\kwd{alternative poly-adenylation}
\kwd{alternative splicing}
\kwd{differential exon usage}
\kwd{UTR}
\kwd{untranslated region}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

Coding sequences in eukaryotic mRNAs are generally flanked by transcribed but untranslated regions (UTRs) which can impact RNA stability, translation, and localization \cite{Lewis1995TheMetabolism}. In particular, the length of 3' UTRs often varies even within a given gene due to the use of different poly-adenylation (polyA) sites \cite{Tian2016AlternativePrecursors}, leading especially to the inclusion or not of regulatory elements such as binding sites for microRNAs (miRNAs) or RNA-binding proteins \cite{Fabian2010RegulationMicroRNAs}. Alternative poly-adenylation (APA) is highly prevalent in mammals \cite{Derti2012AMammals} and has been shown to be important to a variety of biological phenomena \cite{Sandberg2008ProliferatingSites,Mayr2009WidespreadCells,Miura2013WidespreadBrain, Ha2018QAPA:Data}.

A number of methods for 3' end sequencing have been developed with the goal to map APA sites \cite{Fox-Walsh2011AFormation,Fu2011DifferentialSequencing,Zheng20163READS+RNA,Jan2011Formation3UTRs,Shepard2011ComplexPAS-Seq,Derti2012AMammals,Hwang2017CTag-PAPERCLIPActivation}, leading to the development of atlases such as \textit{PolyASite} \cite{Herrmann2020PolyASiteSequencing} or \textit{PolyA\_DB} \cite{WangPolyadb2018}. As such methods are only marginally used, however, it would be beneficial to leverage the widespread availability of traditional RNA-seq for the purpose of identifying changes in 3' UTR usage. A chief difficulty here is that most UTR variants are not catalogued in standard transcript annotations, limiting the utility of standard transcript-level quantification based on reference transcripts, such as \textit{salmon} \cite{PatroSalmon2017}. Nevertheless, a number of methods have been developed to this purpose. Methods like \textit{DaPars} \cite{Xia2014DynamicTypes} and \textit{APAtrap} \cite{Ye2018APAtrap:Data} try to infer new polyA sites from read coverage changes from RNA-seq experiments, however the depletion of RNAseq coverage at the 3' end of transcripts makes the precise inference of polyA sites challenging \cite{Wang2009RNA-Seq:Transcriptomics}. Other tools like \textit{QAPA} \cite{Ha2018QAPA:Data} and \textit{APAlyzer} \cite{Wang2020APAlyzer:Isoforms} use already available polyA site databases but only compare the usage of the most proximal polyA sites to distal ones in a pairwise fashion and fail to grasp to full complexity of dynamic APA when there are three or more polyA sites, which is the case for approximately half of mammalian transcripts \cite{Derti2012AMammals}. Furthermore they do not make use of the already proven statistical tools to analyse different exon usage (DEU) from count data \cite{Anders2012DetectingData,Robinson2009EdgeR:Data,Law2014Voom:Counts,Ritchie2015LimmaStudies}. These tools take into account the inherent properties of read count distributions and are arguably more appropriate to analyse differences in relative polyA site usage, which is conceptually highly similar to DEU. We therefore developed \textit{diffUTR}, which streamlines and improves upon well established DEU tools, and leverages them, along with polyA site databases, to infer alternative 3' UTR usage across conditions.

\section*{Results}
\subsection*{Streamlining differential bin/exon usage analysis}

Popular bin-based DEU methods are provided by the \textit{limma} \cite{Ritchie2015LimmaStudies, Law2014Voom:Counts}, \textit{edgeR} \cite{Robinson2009EdgeR:Data} and \textit{DEXSeq} \cite{Anders2012DetectingData} packages. However, their usage is not straightforward for non-experienced users, and their results often difficult to interpret. We therefore developed a simple workflow (Figure \ref{fig:scheme}A), usable with any of the three methods but standardizing inputs and outputs. In particular, bin annotation and quantification, as well as different usage results, are all stored in a \texttt{RangedSummarizedExperiment} \cite{Morgan2018SummarizedExperiment:Container}, which facilitates data storage and exploration, and enables advanced plotting functions irrespective of the underlying method. \textit{diffUTR} is flexible in its application, and supports the use of strand information if available.

\begin{figure}[h]
\includegraphics[width=0.8\textwidth]{figure1.pdf}
\caption{\textbf{Overview. A:} \textit{diffUTR} workflow. Bins are prepared from various types of gene annotations as well as, optionally, additional APA-driven segmentation and extension, then read counts within bins as well as bin information are stored in a standardized \texttt{RangedSummarizedExperiment}, which can then be used as an input for any of the three DEU methods, producing again a standardized output that can be used with the package's plotting functions. \textbf{B:} Schematic of bin preparation. APA sites are used to further segment and extend disjoined gene bins.}
\label{fig:scheme}
\end{figure}

\subsection*{Improvement to \textit{diffSplice}}
\textit{diffUTR} also implements an improved version of \textit{limma}'s \texttt{diffSplice} method which does not to assume constant residual variance across bins of the same gene (see \hyperref[sec:diffSplice2]{diffSplice2}). To test the effect of these modifications in a standard DEU setting, we ran both versions (as well as the other two DEU methods) on simulated data from a previous DEU benchmark \cite{Soneson2016IsoformUsage}. The precision and recall results (Figure \ref{fig:pr1}A) confirmed the previously observed superiority of \textit{DEXSeq} and, more generally, the imperfect false discovery rate (FDR) control. Importantly, it also confirmed that our improved \texttt{diffSplice2} method outperforms the original, at no additional computing cost.

\begin{figure}[h]
\includegraphics[width=0.98\textwidth]{figure2.pdf}
\caption{\textbf{FDR and recall (TPR) on simulated data. A:} In the classical DEU context. \textbf{B:} In the differential UTR usage context. The dashed line indicates a real False Discovery Rate (FDR) of 5\%, and the dots indicate nominal FDRs of 10, 5 and 1\%. \textit{diffUTR} methods far outperform \textit{QAPA} and \textit{DaPars}. In both contexts, our modifications to \texttt{diffSplice} significantly improve its performance.}
\label{fig:pr1}
\end{figure}

\subsection*{Application to differential UTR usage and benchmark on a simulation}

We next sought to evaluate the methods when applied for differential UTR analysis. For this purpose, APA sites are used to further segment and extend UTR bins, as illustrated in Figure \ref{fig:scheme}B (see methods for the details). Given the absence of RNAseq data with a differential UTR usage ground truth, we simulated reads with known UTR differences from real data (see \hyperref[sec:sim]{Simulated Data}). We then ran the different \textit{diffUTR} methods (as well as the unmodified \texttt{diffSplice} variant), and compared them to alternative methods. While \textit{DaPars} \cite{Xia2014DynamicTypes} and \textit{APAlyzer} provide gene-level significance testing, \textit{QAPA} \cite{Ha2018QAPA:Data} does not, and our attempts to use its equivalence classes with standard transcript usage methods (see methods) gave very poor results. Therefore, for the purpose of comparison we tried to alternatives: simply ranked genes according to \textit{QAPA}'s main output, i.e. the absolute difference in polyA site usage between conditions ($|\Delta PAU|$), labeled in \ref{fig:pr1}B as \textit{QAPA.dPau}, or running \textit{t}-tests on the log-transformed PAU values, labeled as \textit{QAPA.qval}. Since \textit{APAlyzer} produces different analyses for genes' 3' end and intronic APA usage, we used both the 3' end results and a combination of the two (the latter shown as \textit{APAlyzer2}). As Figure \ref{fig:pr1}B shows, all \textit{diffUTR} methods outperformed alternatives by far. On this test, our improved \texttt{diffSplice2} nearly equaled \textit{DEXSeq}'s performance, at a fraction of the computing costs.

\subsection*{Differential UTR usage in real data}

We next sought to tests \textit{diffUTR} in real data. First, since 3' UTRs are known to generally lengthen during neuronal differentiation \cite{Blair2017WidespreadDifferentiation,Ha2018QAPA:Data}, we expected to observe a skew towards positive fold changes of 3' UTR bins when comparing RNAseq experiments from embryonic stem cells (ESC) and ESC-derived neurons. We therefore re-analyzed data from \cite{WhippleImprinted2020} and observed clearly the expect skew among statistically-significant genes, especially for bins with a higher expression (Figure \ref{fig:realdata}A).

We next found both 3' sequencing and standard RNAseq data from samples of mouse hippocampal slices undergoing Forskolin-induced long-term potentiation \cite{Fontes2017Activity-DependentPotentiation}, which enabled us to use the 3' sequencing data as a truth for analysis performed on the standard RNAseq (Figure \ref{fig:realdata}B and Supplementary Figure 1). In this case we represent the results through Receiver-operator characteristic (ROC) curves since the Precision-recall curves make the differences less visible due to the lower general power. Although power to detect UTR changes is necessarily low with respect to 3' sequencing, we again observed that \textit{diffUTR} methods clearly outperformed all alternative methods.

\begin{figure}
\includegraphics[width=0.97\textwidth]{figure3.pdf}
\caption{\textbf{Differential UTR analysis on real data. A:}. 3' UTR lengthening during neuronal differentiation. Plotted are the UTR bins found statistically significant (bin- and gene-level FDR both < 0.1) by \textit{diffUTR} (diffSplice2) when comparing in vitro differentiated neurons to mouse embryonic stem cells. The color indicates the point density. The clear skew towards a positive bin-level foldchange (indicative, in most cases, of a UTR lengthening), especially for bins with a higher mean count (CPM=counts per million reads sequenced). \textbf{B:} Receiver-operator characteristic (ROC) curves of differential UTR usage analysis on the LTP dataset, using 3' sequencing to establish the ground truth. The axes are square-root-transformed to improve visibility, and only a subset of method variations are shown (see Supplementary Figure 1 for all variants).}
\label{fig:realdata}
\end{figure}


\subsection*{Exploring differential exon/UTR usage results}

\textit{diffUTR} provides three main forms of plot to explore differential bin usage analyses, each with a number of variations. Figure \ref{fig:plots} showcases them in the context of long-term potentiation of mouse hippocampal neurons \cite{Fontes2017Activity-DependentPotentiation}. \texttt{plotTopGenes} (Figure \ref{fig:plots}A) provides gene-level statistic plots (similar to a `volcano' plot), which come in two variations. For standard DEU analysis, absolute bin-level coefficients are weighted by significance and averaged to produce gene-level estimates of effect sizes. For differential 3' UTR usage, where bins are expected to have consistent directions (i.e. lengthening or shortening of the UTR) and where their size is expected to have a strong impact on biological function, the signed bin-level coefficients are weighted both by size and significance to produce gene-level estimates of effect sizes. By default, the size of the points reflects the relative expression of the genes, and the colour the relative expression of the significant bins with respect to the gene.

\texttt{deuBinPlot} (Figure \ref{fig:plots}B) provides bin-level statistic plots for a given gene, similar to those produced by \textit{DEXSeq} and \textit{limma}, but offering more flexibility. They can be plotted as overall bin statistics, per condition, or per sample, and can plot various types of values. Importantly, since all data and annotation is contained in the object, these can easily be included in the plots. Figure \ref{fig:plots}B shows a lengthening of the Jund 3' UTR in the LTP group.

Finally, \texttt{geneBinHeatmap} (Figure \ref{fig:plots}C) provides a compact, bin-per-sample heatmap representation of a gene, allowing the simultaneous visualization of various information. We found these representations particularly useful to prioritize candidates from differential bin usage analyses. For example, many genes show differential usage of bins which are generally not included in most transcripts of that gene (low count density), and are therefore less likely to be relevant. 

\begin{figure}
\includegraphics[height=0.6\textheight]{examples.pdf}
\caption{\textbf{Plotting functions. A:} \texttt{plotTopGenes} provides significance and effect size statistics aggregated at the gene level. \textbf{B:} \texttt{deuBinPlot} provides a more flexible version of the bin-level gene plots generated by common DEU packages. Shown here is the upregulation of Jund 3' UTR upon LTP. \textbf{C:} \texttt{geneBinHeatmap} provides a compact, bin-per-sample heatmap representation of a gene.}
\label{fig:plots}
\end{figure}

\subsection*{Further variations tested}

During implementation, we tested other changes to the method which were ultimately discarded as they did not improve performance, but which we here briefly report.

First, differential UTR analysis differs from typical differential exon usage analysis in that in that the vast majority of UTR bins are consecutively transcribed, meaning that changes in the usage of a bin should also be visible in downstream bins. We therefore reasoned that it would be beneficial to use this property to improve statistical analysis. We reasoned that connected bins with significant fold changes in the same direction could be unified and their p-values aggregated, and tested a rudimentary implementation using Fisher's aggregation. However, this decreased accuracy and led to a worse FDR control (Supplementary Figure 2).

Second, most methods compare bin-level foldchanges to gene-level ones to identify, and we reasoned that, especially for genes with more UTR bins than CDS bins, including counts of 3' UTR when calculating overall gene expression could underestimate the gene expression and possibly mistake the UTR foldchange for the gene foldchange. We therefore tried a modification of \textit{diffSplice} to only calculate the gene foldchange from coding sequence (CDS) bins and then compare it to the individual bins. Again, this approach proved unsuccessful (Supplementary Figure 3).

\section*{Discussion}

\textit{diffUTR} streamlines DEU analysis and outperforms \textit{DaPars} and \textit{QAPA} alternative methods in inferring UTR changes, which demonstrates the utility of harnessing powerful, well-established frameworks for new ends. It must be noted that the way in which the simulation was performed, i.e. elongating transcripts to the next polyA site(s), is similar to the way \textit{diffUTR} disjoins the annotation into bins, which could cause a bias towards this method (as well as \textit{QAPA}, which also makes use of alternative polyA sites). However, \textit{diffUTR} outperformed these methods is so considerable a fashion that it is highly unlikely to be only a product of such bias. Nevertheless, this highlights the need for real, high quality datasets where at least partial ground truths are available for benchmarking purposes.

Similar to DEU tools \cite{Soneson2016IsoformUsage}, \textit{diffUTR} fails to control the FDR correctly, and our attempts so far to improve this remained unsuccessful. In addition, in contrast to DEU, where exons are subject to splicing in a potentially independent fashion, 3' UTRs typically do not undergo splicing and therefore only differ in lengths between condition. This means that the behavior of a UTR bin is dependent on that of upstream bins, a property which could be exploited to improve accuracy. However, our simple attempt to do so by combining p-values of consecutive bins did not have the desired outcome, pointing to the need of more research in this direction.

Further, the bin-based approach has the drawback of not pinpointing the exact UTR locations: it is limited to the bin resolution, and the bins themselves are limited by incomplete transcript and APA annotations. Additionally, because there is a significant drop off in read coverage at the end of transcripts, we have observed that it is often bins upstream of the actual UTR lengthening/shortening event which give a statistically-significant signal rather than the one truly affected. This is why we have provided tools to enable the further inspection of events in a given gene.

Finally, the results of bin-based analyses are limited by the overlaps between genes, an issue on which differential transcript usage analysis approaches appear superior (e.g. \cite{Tiberi2020BANDITS:Uncertainty}). However, transcript usage analysis tools are dependent on the completeness of the transcript annotation, while bin-based approaches are more open to the discovery of unannotated transcript variants, which is especially relevant for differential UTR usage. While \textit{DEXSeq} remains the tool of predilection for such analyses, it scales very badly to larger sample sizes, and alternatives might be needed in some contexts. Our changes to \textit{limma}'s original \texttt{diffSplice} method consistently result in more accurate predictions, making this new method the best compromise for bin-based approaches when \textit{DEXSeq} is not applicable. More generally, it also shows that even with well-established approaches, there is still room for incremental, but non-negligible improvement.


\section*{Methods}

The data objects and code used to produce the figures are available through the \url{https://github.com/plger/diffUTR_paper} repository.

\subsection{RNAseq data processing}

For the evaluation of \texttt{diffSplice2} in a standard DEU case, we used bin count data obtained from the authors of the original DEU benchmark \cite{Soneson2016IsoformUsage}. For other datasets, reads were downloaded from the SRA, aligned to the GRCm38.p6 genome using STAR 2.7.3a with default parameters and the GENCODE M25 annotation as guide. The same gene annotation was used as input for bin creation.

\subsection{diffUTR}

\textit{diffUTR} is implemented as a Bioconductor package making use of the extensive libraries available, especially the \textit{GenomicRanges} package \cite{Lawrence2013SoftwareRanges} and the different DEU methods (see \hyperref[sec:DE]{Differential analysis}). Its source code is available at \url{https://github.com/ETHZ-INS/diffUTR}.

\subsubsection{Preparing bins}
\label{sec:bins}

Exons are extracted from the genome annotation and flattened into non-overlapping bins (Figure \ref{fig:scheme}B). In other words, the exon annotation is fragmented into the widest ranges where the set of overlapping features is the same. Bins that do not overlap with coding sequences (CDS) and belong to a protein coding transcript are labeled as UTR and the rest as CDS. When APA sites are also provided as input (for the purpose of this article, polyAsite v2.0 sites were used), bins are further segmented and/or extended. For this the closest upstream CDS or UTR is found for every poly(A) site and the UTR is defined from this boundary to the polyA site and assigned to the corresponding gene and transcript (Figure \ref{fig:scheme}B). If the newly defined UTRs exceeds a predefined length specified by \texttt{maxUTRbinSize} (default is 15000bp), it is ignored as unlikely to be a real UTR. Moreover, if the start of a gene is the closest upstream sequence before any UTR or CDS the newly defined UTR is ignored to avoid assignment problems. In order to later differentiate between regions that are 3' UTR and other kind of UTR, regions that are downstream of the last CDS of a given transcript were labeled as 3' UTR. The label `non-coding' is assigned to all bins that have no protein coding transcript overlapping it.

If a bin originates from regions belonging to different genes, the bin is duplicated and assigned once to each gene, so that every gene contains the same fragment once. Alternatively, the \texttt{genewise} argument can be used so that only exons belonging to the same gene are considered when flattening.

\subsubsection{Quantification}

For quantification, \texttt{countFeatures()} uses the \texttt{featureCounts()} function from the \textit{Rsubread} package \cite{Liao2014FeatureCounts:Features} to count previously mapped reads overlapping each bin. By default every read is assigned once to every bin it overlaps with and can therefore be counted multiple times, which is needed because many bins are shorter than the read length. Alternative counting methods, such as \texttt{summarizeOverlaps()} from the \textit{GenomicAlignments} package \cite{Lawrence2013SoftwareRanges} performed considerably worse. The function returns a \texttt{RangedSummarizedExperiment} object \cite{Morgan2018SummarizedExperiment:Container}, containing the read counts as well as the bin annotation. 

\subsubsection{Differential analysis}
\label{sec:DE}
Three wrappers implement corresponding DEU methods on the \linebreak \texttt{RangedSummarizedExperiment} object previously generated, returning results as further standardized annotation within the object. For the differential UTR analysis there are 3 different wrapper methods provided, which work with the previously returned SummarizedExperiment object. For differential UTR analysis, gene-level results are obtained by filtering the bin-level results for those assigned to the type UTR and or 3' UTR, and setting all other p-values to 1 before aggregation.

\paragraph{diffSpliceDGE.wrapper()}

This is a wrapper around \textit{edgeR}'s DEU method based on fitting a negative binomial generalized linear model \cite{Robinson2009EdgeR:Data}. In a first step the bins are filtered to decide which have a large enough read counts to be kept for the statistical analysis (\texttt{filterByExpr()}), the library sized are normalized (\texttt{calcNormFactors()}) and the dispersion is estimated (\texttt{estimateDisp()}). After this the model is fitted (\texttt{glmFit()}). If the option \texttt{QLF = TRUE} (default), an extended model is fitted, using quasi-likelihood methods to account for gene specific variability (\texttt{glmQLFit()}). In the last step bin fold changes are tested to be different from overall gene fold changes, using a likelihood ratio test or a quasi-likelihood F-Test depending on the \texttt{QLF} option chosen (\texttt{diffSpliceDGE()}).  The gene level p-values are obtained by the Simes' method \cite{Simes1986AnSignificance}. 
\paragraph{DEXseq.wrapper()}

In this method the standard \textit{DEXseq} differential exon usage pipeline \cite{Anders2012DetectingData} is implemented. It is similarly to edgeR based on fitting a negative binomial model but instead of comparing fold change differences between bins and genes, \textit{DEXseq} compares a full model containing a term corresponding to the change in exon usage between conditions to a reduced model without this term. The two fits are compared using a $\chi^2$ likelihood-ratio test. The libraries are normalized (\texttt{estimateSizeFactor()}), the dispersion is estimated (\texttt{estimateDispersion()} and the models are fitted (\texttt{testForDEU()}). In a last step the fold changes between the bins are estimated ( \texttt{estimateExonFoldChanges()}). To obtain gene level results the function \texttt{perGeneQValue()} is used, which is based on the Šidák method \cite{Sidak1967RectangularDistributions}.

\paragraph{diffSplice.wrapper() and diffSplice2}
\label{sec:diffSplice2}
This method implements the differential exon usage pipeline of \textit{limma} for RNA-seq data \cite{Ritchie2015LimmaStudies}. The pre-processing is identical to \texttt{diffSpliceDGE.wrapper()}, then the precision weights are estimated with (\texttt{limma::voom()}) and the linear models are fitted (\texttt{limma::lmFit()}). In the last step, bin fold changes are tested to be different from overall gene fold changes, using a moderated t-test (\texttt{diffSplice()} or, by default, \texttt{diffSplice2()} -- see below). The gene level p-values are obtained by the Simes' method \cite{Simes1986AnSignificance}.

The \texttt{diffUTR::diffSplice2} function provides a slightly improved version of \textit{limma}'s original \texttt{diffSplice} method. \texttt{diffSplice} works on the bin-wise coefficient of the linear model which corresponds to the log2 fold changes between conditions. It compares the log2(fold change) of a bin $k$, belonging to gene $g$ called $\hat{\beta}_{k,g}$ to a weighted average of log2(fold change) of all the other bins of the same gene combined $\hat{B}_{k,g}$ (the subscript $g$ will be henceforth omitted for ease of reading). The weighted average of all the other bins in the same gene is calculated by
\begin{equation}
    \hat{B}_{k}= \frac{\sum_{i, i\neq k}^{N}{w_{i} \hat{\beta}_{i}}}{\sum_{i, i\neq k}^{N}w_{i}}
\end{equation}
where $w_{i}=\frac{1}{u_{i}^2}$ and $u_{i}$ refers to the diagonal elements of the unscaled covariance matrix $(X^TVX)^{-1}$. $X$ is the design matrix and $V$ corresponds to the weight matrix estimated by \texttt{voom}. The differences of log2 fold changes, which is also the coefficient returned by \texttt{diffSplice()} is then calculated by $\hat{C}_{k} = \hat{\beta}_{k} - \hat{B}_{k}$. Instead of calculating the t-statistic with $\hat{C}_{k}$, this value is scaled again in the original code:

\begin{equation}
    \hat{D}_{k} =\hat{C}_{k}\sqrt{1-\frac{{w}_{i}}{\sum_{i}^{N}w_{i}}}
\end{equation}

and the \textit{t}-statistic is calculated as:

\begin{equation}
    t_{k} =\frac{\hat{D}_{k}}{u_{k}s_{g}}
\end{equation}

$s_{g}^2$ refers to the posterior residual variance of gene $g$, which is calculated by averaging the sample values of the residual variances of all the bins in the gene, and then squeezing these residual variances of all genes using empirical Bayes method. This assumes that the residual variance is constant across all bins of the same gene. 

In \texttt{diffSplice2()}, we applied three changes to the above method. First, the residual variances are not assumed to be constant across all bins of the same gene. This results in the sample values of the residual variances of every bin now being squeezed using empirical Bayes method, resulting in posterior variances $s_{i}^2$ for every individual bin $i$. Second, the weights $w_{i}$, used to calculate $\hat{B}_{k}$, now incorporate the individual variances by $w_{i}=\frac{1}{s_{i}^2u_{i}^2}$. Third, the $\hat{C}_{k}$ value is directly used to calculate the \textit{t}-statistic, which after all these changes now corresponds to

\begin{equation}
    t_{k} =\frac{\hat{C}_{k}}{u_{k}s_{i}}.
\end{equation}

\subsection{Simulated Data}
\label{sec:sim}
The simulation was done using the \textit{Polyester} R package  \cite{Frazee2015Polyester:Expression} using parameters obtained from the control samples of mouse hippocampus RNAseq \cite{Fontes2017Activity-DependentPotentiation}. Using \textit{salmon} \cite{PatroSalmon2017} with a decoy-aware transcriptome index for the mm10 genome from \cite{Stolarczyk2020Refgenie:Manager}, the abundances for each transcript were first estimated to learn parameters for the simulation. 1000 transcripts from different genes were randomly chosen. The last exon of all these transcripts was lengthened to the next, second next or third next downstream APA site annotated in the polyAsite database \cite{Herrmann2020PolyASiteSequencing}. Duplicates of these transcripts were generated, which had less or no lengthening of their last exon, generating pairs of transcripts with different UTR lengths. For each transcript pair, one transcript was up and the other one down regulated by the same sampled fold change between 1.3 and 5. To make it more realistic, fold changes were also assigned to 300 genes from the set with differential UTR, and 300 genes that did not have differences in UTR usage. Reads were then generated for two conditions with three replicates each using the \texttt{simulate\_experiment()} function with the options \texttt{paired = FALSE},  \texttt{error\_model = "illumina5"}, 
\texttt{bias = "cdnaf"} and \texttt{strand\_specific = TRUE}. The simulated reads are available on figshare at \url{https://dx.doi.org/10.6084/m9.figshare.13726143}.

\subsection{3'-seq analysis}

To establish a set of true relative differences in UTR usage from the 3' sequencing data \cite{Fontes2017Activity-DependentPotentiation}, we downloaded the authors' counts per cluster from the Gene Expression Omnibus (file \texttt{GSE84643\textunderscore 3READS\textunderscore count\textunderscore table.txt.gz}). We used the 3h treatment because we observed it to have the strongest signal, and excluded one sample (A6) that appeared like a strong outlier based on PCA and MDS plots. We kept only clusters with at least 50 reads in at least 2 samples, and used \textit{DEXSeq} to fit a negative binomial on each gene and estimate the significance of the \texttt{cluster:condition} term. We considered as true positives genes with a gene-level and bin-level q-value $\leq$ 0.1, and true negatives genes with a gene-level q-value $\geq$ 0.8. Genes for which all tested methods produced a p-value of 1 or NA (i.e. genes filtered out as too lowly expressed in the standard RNAseq) were excluded for the benchmark.

\subsection{Comparisons with alternatives}

For the comparison of methods, all functions were used with their default parameters and  run according to their manual. As \textit{QAPA} and \textit{DaPars} do not provide means to aggregate the results to a gene level this was implemented separately. For \textit{DaPars} the p-value were aggregated to the gene level by using Simes' method \cite{Simes1986AnSignificance} for comparability with \textit{diffUTR}. Aggregation by taking the minimum p-value of all the transcripts in a gene produced extremely similar results. For \textit{QAPA} $|\Delta PAU|$ was calculated and aggregated to a gene level by taking the maximum from all transcripts of a gene and the genes were ranked by this value. Alternatively, we also tested applying a \textit{t}-test on the log-transformed $PAU$ values (log-transforming had a negligible effect), followed by Simes' gene-level aggregation. Attempts to complement \textit{QAPA} with p-values estimated from established statistical tests working with its equivalence classes, such as BANDITS \cite{Tiberi2020BANDITS:Uncertainty}, did not improve the results and were therefore discarded so as not to distort the original method. Finally, for \textit{APAlyzer2} we combined the 3' UTR and intronic APA analyses by using the minimum of the two p-values. See the \url{https://github.com/plger/diffUTR_paper} repository for details.

We used the following software versions for comparisons: \textit{Polyester} 1.24.0, \textit{DEXSeq} 1.34.0, \textit{edgeR} 3.30.0, \textit{limma} 3.44.0, \textit{DaPars} 0.9.1, \textit{APAlyzer} 1.5.5. For \textit{QAPA}, we used \textit{salmon} 1.3.0 with \texttt{validateMappings}.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare no competing interests beside being the developers of the described package.

\section*{Author's contributions}
  SG developed the bin preparation and the diffSplice modification, and ran most of the analyses. PLG and SG wrote the package and paper. PLG and GS supervised the project.

\section*{Acknowledgements}
  SG performed this research as part of his bachelor thesis in the Interdisciplinary Sciences program at ETH. PLG's position is co-funded by Prof. Mark Robinson (Institute of Molecular Life Sciences, University of Zurich) and Professors Gerhard Schratt, Johannes Bohacek and Isabelle Mansuy (Institute of Neuroscience, ETH Zurich). GS is supported by grants from the SNF (SNF\_179651, SNF\_189486) and the ETH (ETH-24 18-2 (NeuroSno)). We thank the Robinson group (UZH) for feedback.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc_article}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\iffalse

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      A short description of the figure content
      should go here.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
\caption{Sample table title. This is where the description of the table should go.}
      \begin{tabular}{cccc}
        \hline
           & B1  &B2   & B3\\ \hline
        A1 & 0.1 & 0.2 & 0.3\\
        A2 & ... & ..  & .\\
        A3 & ..  & .   & .\\ \hline
      \end{tabular}
\end{table}
\fi
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Supplementary Figures}
    Supplementary Figures 1-3.

\end{backmatter}
\end{document}
